llvm/test/MC/Sparc/sparcv9-instructions.s

```diff
@@ -557,6 +557,36 @@
         ! V9: prefetch  [%i1+%i2], #one_read  ! encoding: [0xc3,0x6e,0x40,0x1a]
         prefetch  [ %i1 + %i2 ], #one_read
 
+        ! V8:      error: malformed ASI tag, must be a constant integer expression
```

Comment:
The error message is wrong.
I guess parseASITag / parsePrefetchTag should return NoMatch on no match rather than issue an error.

Reply:
It is because parseAbsoluteExpression fails. When it fails, we're no longer allowed to return NoMatch as the method consumes tokens. It should not be called at all if the tokens to be parsed do not look like an expression.
I can suggest two ways to overcome this:
* Restrict numerical tags to be integer literals and do not use parseAbsoluteExpression. I recall you saying that the numerical tag can indeed be an expression, so this might not be the way to go.
* Check that the next token can be a start of an expression before calling parseAbsoluteExpression. Those should be Identifier, Integer, Plus, Minus, LParen, and possibly Tilde and Exclaim. Such kind of check is performed in, e.g., parseCallTarget and parseSparcAsmOperand.

Reply:
With the above suggestion it should now yield an "unexpected token" error.
If you wish to get "unavailable in V8" error, you need to pass ParseForAllFeatures == true to MatchOperandParserImpl so that parsePrefetchTag is called in V8 mode too. This will probably result in many test changes and should be done separately.

Reply:
I hoped a fix would be trivial. Let's do it in separate patch then.